### ISSUE TITLE
application crash with "corrupted double-linked list"

### ISSUE THREAD

**Issue by bionick7** [godotengine/godot#67263]
### Godot version

4.0 beta 2 mono

### System information

Ubuntu 22.04.1 / GTX 960 / Vulkan

### Issue description

Godot application (not editor) crashes with either "corrupted double-linked list" or "malloc(): smallbin double linked list".

Here is the output from the console (2 different occurrences)

`Godot Engine v4.0.beta2.mono.official.f8745f2f7 - https://godotengine.org
Vulkan API 1.2.0 - Using Vulkan Device #0: NVIDIA - NVIDIA GeForce GTX 960
libdbus-1.so: cannot open shared object file: No such file or directory
libdbus-1.so: cannot open shared object file: No such file or directory
 
ERROR: Condition "!d.has("handle_modes")" is true. Returning: false
   at: _set (scene/resources/animation.cpp:319)
ERROR: Condition "!d.has("handle_modes")" is true. Returning: false
   at: _set (scene/resources/animation.cpp:319)
WARNING: res://Scripted agents/Systems3D/Guns/Gettling/Gettling.tscn:4 - ext_resource, invalid UUID: uid://dabx56bsjisvu - using text path instead: res://Scripted agents/Systems3D/Guns/Gettling/GettlingTurret.tscn
     at: load (scene/resources/resource_format_text.cpp:448)
WARNING: res://Scripted agents/Systems3D/Guns/Gettling/GettlingTurret.tscn:10 - ext_resource, invalid UUID: uid://6m0p06s5bt5v - using text path instead: res://Graphics/materials/exhaust_process_shader.tres
     at: load (scene/resources/resource_format_text.cpp:448)
malloc(): smallbin double linked list corrupted
Aborted (core dumped)`

sometimes that second last line is `corrupted double-linked list`

### Steps to reproduce

Despite how often it happens, the bug is hard to reproduce and happens at will It seams however, that it always triggers on Signal Emission in C#. I have also tried and failed to reproduce this in a smaller project.

**Comment by Seikho** [comment]
I also frequently experience the same without without the preceding output.
My application simply exits and the only log I get is `corrupted double-linked list` and sometimes I simply get no output at all. I've encountered with with C# beta 1, beta 2, and beta 3. I'm usually on Ubuntu 22.04.1, but have also encountered the issue on Windows 11.
I'm actively trying to narrow down the cause, but haven't been able to.

**Comment by eumario** [comment]
I have also ran into this problem, on Manjaro Linux, with crashes that occur at random.  There is no specific "here's where it is crashing", I could be doing one sequence, and it'll crash, do the same sequence in the next run, no issue.  Tried everything from creating and adding nodes on fly, to keeping nodes stored in a dictionary and only add/remove them when switching.  Sometimes it will even just crash in the middle, with nothing going on but it sitting and idly run.

**Comment by Dunkhan** [comment]
I have a reproduction, it is not straightforward but it should be better than nothing.

1. Download the project branch https://gitlab.com/gametheatre/poriferareader/Porifera/-/tree/bug_67263
2. After the list populates, click the 'Republish' button on the right side of any item
3. On the resulting popup, click the 'TestPublish' checkbox
4. click another Republish button on another item without closing the popup
5. repeat steps 3 - 4 rapidly until crash, usually happens with 5 attempts

The error message 'corrupted double linked list' only shows very rarely, usually there is no error message at all. It stands to reason that the cause of the crash is the same every time though.

n.b. Please don't judge my application based on this branch, I have deliberately disabled most of the functionality to make this reproduction simpler.

**Comment by bionick7** [comment]
>The error message 'corrupted double linked list' only shows very rarely, usually there is no error message at all.

I only got the error message when running the program from the console, but then I get it consistently.

**Comment by bionick7** [comment]
Also @Dunkhan, I could not access your reproduction project. It's probably private. Maybe just upload the zip instead

**Comment by Dunkhan** [comment]
So sorry, I am new to gitlab thought public was the default setting. It should be working now. Zip is too large for github but I can upload one somewhere else if that is preferred

**Comment by bionick7** [comment]
I coudn't find any of the buttons you describe in your reproduction project. I think your app needs some preexisting data to work properly
![image](https://user-images.githubusercontent.com/42216907/197850036-12196352-663c-4e02-974c-e06b8b1b18fe.png)
![image](https://user-images.githubusercontent.com/42216907/197850066-7f088e52-f653-4d45-85c4-f2000594b115.png)
It is interesting to see though that this also happens in Godot3.x

**Comment by Seikho** [comment]
Ah, well that's a bummer that it happens in 3.x as well. I've been porting my work to 3.5 to get past the crash. At least I can give up on porting back to 3.5.1 then.

After going through the source code of the repro here, I'm not seeing too much in common with my code that causes the same problem. I'm not using any Tasks, `async/await`, 3rd party dependencies, custom signals/events, etc. I've removed all `static` references from my project as well and that didn't fix the issue either.
This project isn't dynamically adding/removing scenes like I am either so it probably isn't that either.
The only thing I see in common is the use of dictionaries.

**Comment by Seikho** [comment]
After digging deeper, my issue went away once I removed the only invocation of `RPC`. The repro posted by @Dunkhan also contains a network call.
I've managed to create a minimal reproduction at https://github.com/seikho/godot-crash-repro.

The crash an at seemingly "random" which leads me to believe it could be related to #67563 as well.

**Comment by eumario** [comment]
After seeing @Dunkhan  having the same error in 3.5.1, and me running into a similar issue, I went down a rabbit hole, and think I have possibly tracked the problem.  This is a theory, as I have nothing to fully prove this to be the case, but in my particular case, I'm coding in Godot 4.0 with TwitchLib and OBS-WebSocket-Net, when I ran into this error, I backed to 3.5.1, and was making progress with my development, till it started crashing randomly out of no where.  It would take multiple attempts to cause the error to popup, not like with Godot 4, where it would happen almost immediately.  But when I received a signal that OBS was changing scenes, I would change the scene in Godot as well.  On 4.0, this caused the corrupted double-linked list error.  When backported to 3.5, it would take a few changes to the scene, and would randomly crash on the Scene changed.  I tested with Keyboard input in Godot to see if just changing the scenes was the cause of the issue, and when I did it with the keyboard, it had no issue, it would change scenes no matter what, not matter how slow or fast I would do it, the program would never crash.

Then I got the idea from @Dunkhan, to use CallDeferred.  In the 3.5.1 project, it instantly solved the problem, I would get no further crashes, but in the 4.0, it would stabilize the program, and wouldn't instantly crash.  Eventually, it would crash, so it made me think that maybe there was a Thread issue going on, where an attempt to free memory was happening, that caused the a Race Condition, as no proper mutex locking was occurring.  Then it reminded me of the other two Memory bugs that I saw on the tracker in regards to 4.0, #67460 and #67563.  Both of these are similar issues where Godot is crashing when attempting to free memory.  It may be, that we need in the backend a way to lock memory during a GC Collection, to ensure that no cross-thread shenanigans are causing the memory corruption from occurring.

I will make a reproduction of the issue with Godot 4.0 and 3.5.1, using just OBS-WebSocket-Net, though it will require having OBS Studio installed with OBS-WebSocket setup to test, but may be able to give valuable information.

**Comment by Dunkhan** [comment]
> I coudn't find any of the buttons you describe in your reproduction project. I think your app needs some preexisting data to work properly

Did you use the branch I linked? I added the lines
```
this.incomingFeedUrls.Add("https://www.reddit.com/r/godot/.rss");
this.outgoingFeeds.Add(new () {Title = "TestPublish"});
```
specifically to make the repro work

**Comment by bionick7** [comment]
My bad, I was on the main branch, made in 3.5. The new one is working as described.
> After digging deeper, my issue went away once I removed the only invocation of RPC. The repro posted by @Dunkhan also contains a network call.

It is maybe worth mentioning that I have make no network calls whatsoever. For me, the issue goes away (in one part at least) after i stop emitting a signal. I don't explicitly do threading, but signals handling might be threaded

**Comment by eumario** [comment]
A recent test of Beta 4 shows that in my main project, it isn't crashing anymore with the memory corruption error, however, it just locks up, and won't continue further.

**Comment by Dunkhan** [comment]
Who here is using RichTextLabel in their interface?
I tried to create a minimal reproduction to see where my project was going wrong and maybe help shed some light on this bug and the key that broke everything was RichTextLabels being present.

I am going to report this as a separate bug in the meantime, if it turns out to be the same issue I will close it as a duplicate.
I can't get it to show the error message about the double linked list so I think it might be a separate issue

https://github.com/godotengine/godot/issues/68207

Based on the report it may not be obvious why these bugs are related, but in a more complex reproduction I was able to crash the application by getting a signal to manipulate RichTextLabels. Unfortunately I am still not able to reliably reproduce the crash and will report that separately if I manage.

**Comment by eumario** [comment]
I had totally forgotten that I was using RTL in my project, and that there was a change to the way the BBCode was parsing.  They did add a Threaded option to RTL when parsing BBCode, which would explain why there would be what I was thinking was a Thread Racing condition.  This could be very well the culprit.  @Dunkhan, have you tried using Beta 4, to see if it does as I reported, and just locks up, instead of crashing to the memory corruption error, or just locking Godot up completely?

**Comment by Dunkhan** [comment]
I am still not able to get export templates working when I self build, there is some step missing in the guide. I got further but there is still an issue with nuget which the compile guide does not even mention so... I will let you know once beta4 is released

**Comment by eumario** [comment]
Beta 4 is released.  https://downloads.tuxfamily.org/godotengine/4.0/beta4/  They just didn't make any big announcements.

**Comment by bionick7** [comment]
> A recent test of Beta 4 shows that in my main project, it isn't crashing anymore with the memory corruption error, however, it just locks up, and won't continue further.

I get the same issue, although it always freezes directly after loading (_Ready() is still called) instead of crashing randomly. Once again I fail to reproduce the bug with a smaller project.

**Comment by Zireael07** [comment]
@bionick7: you mean the editor locks up on opening the project?

I've had that issue and narrowed it down to reopening scripts. Disabling that editor setting fixed the problem for me

**Comment by bionick7** [comment]
No, I mean my app freezing after launch

**Comment by Dunkhan** [comment]
For me turned out to be due to threads accessing the scenetree in an unsafe way. Someone on discord informed me that the scenetree is not thread safe and should never be accessed outside the main thread. This means my UI updates needed to be deferred. If you are confused about how to do this effectively in C# (this too me a while to find out) the specific solution I used can be found [here](https://www.reddit.com/r/godot/comments/yt3th2/functional_version_of_calldeferred_for_c/)

tl:dr ignore the reproduction I posted, if my solution does not work for other people here then I never had the same issue as them, or at least had it so rarely that I haven't seen it since. At the very least the reproduction does not appear to reflect an ongoing issue (other than perhaps the log messages being non-existent/unhelpful)

**Comment by bionick7** [comment]
> > A recent test of Beta 4 shows that in my main project, it isn't crashing anymore with the memory corruption error, however, it just locks up, and won't continue further.
> 
> I get the same issue, although it always freezes directly after loading (_Ready() is still called) instead of crashing randomly. Once again I fail to reproduce the bug with a smaller project.

So, to clarify, the issue I described here was unrelated to the main issue, which still persists in beta 4 as described in the original post. I neither do multithreding explicitly, nor do I have network calls or Rich Text labels. So most of your issues are probably different, but maybe related, issues to mine.

**Comment by eumario** [comment]
@bionick7 Something I wanted to confirm with you, as I had another hunch in regards to this.  I understand that you are not doing any explicit multi-threading, and you stated that it stopped after you emitted a custom signal.  My question is, are you using any 3rd party library that implements Events specifically for call backs to the library itself?  Just trying to eliminate all possibilities.

**Comment by bionick7** [comment]
I barely have any 3rd party libraries in my project. The Signals I was talking about connected either to my own c# code or vanilla godot methods (Label.set_text on the relevant one)

**Comment by raulsntos** [comment]
I opened PR https://github.com/godotengine/godot/pull/69194 that may fix this issue, can you try it out to confirm?

Linux users can download the artifact built by CI: https://github.com/godotengine/godot/suites/9514119809/artifacts/451640372
For other platforms, you'll have to build Godot yourself.

**Comment by bionick7** [comment]
Yes, that's it. It actually fixes the issue. Amazing :+1:

**Comment by neikeq** [comment]
Re-opening as the PR isn't merged yet :stuck_out_tongue: (it will auto-close when merged)

**Comment by Dunkhan** [comment]
This fixed it for me to. Thanks so much for sorting this.
